oxide-auth-db/src/db_service/redis.rs

```diff
@@ -0,0 +1,152 @@
+
+    /// The registered redirect uri.
+    /// Unlike `additional_redirect_uris`, this is registered as the default redirect uri
+    /// and will be replaced if, for example, no `redirect_uri` is specified in the request parameter.
+    pub redirect_uri: String,
+
+    /// The redirect uris that can be registered in addition to the `redirect_uri`.
+    /// If you want to register multiple redirect uris, register them together with `redirect_uri`.
+    pub additional_redirect_uris: Vec<String>,
+
+    /// The scope the client gets if none was given.
+    pub default_scope: Option<String>,
+
+    /// client_secret, for authentication.
+    pub client_secret: Option<String>,
+}
+
+impl StringfiedEncodedClient {
+    pub fn to_encoded_client(&self) -> EncodedClient {
+        let redirect_uri =  Url::parse(&self.redirect_uri).unwrap();
+        let uris =  &self.additional_redirect_uris;
+        let additional_redirect_uris = uris.into_iter().fold(vec![], |mut us, u| { us.push(Url::parse(u).unwrap()); us});
+        let client_type = match &self.client_secret {
+            None => ClientType::Public,
+            Some(secret) => ClientType::Confidential {passdata: secret.to_owned().into_bytes()},
+        };
+        EncodedClient{
+            client_id: (&self.client_id).parse().unwrap(),
+            redirect_uri,
+            additional_redirect_uris,
+            default_scope: Scope::from_str(self.default_scope.as_ref().unwrap_or(&"".to_string()).as_ref()).unwrap(),
+            encoded_client: client_type
+        }
+    }
+
+    pub fn from_encoded_client(encoded_client: &EncodedClient) -> Self {
+        let additional_redirect_uris = encoded_client.additional_redirect_uris
+            .iter()
+            .map(|u| u.to_owned().into_string())
+            .collect();
+        let default_scope = Some(encoded_client.default_scope.to_string());
+        let client_secret = match &encoded_client.encoded_client{
+            ClientType::Public => None,
+            ClientType::Confidential { passdata} => Some(String::from_utf8(passdata.to_vec()).unwrap())
+        };
+        StringfiedEncodedClient{
+            client_id: encoded_client.client_id.to_owned(),
+            redirect_uri: encoded_client.redirect_uri.to_owned().into_string(),
+            additional_redirect_uris,
+            default_scope,
+            client_secret
+        }
+    }
+}
+
+
+impl RedisDataSource {
+    /// users can regist to redis a custom client struct which can be Serialized and Deserialized.
+    pub fn regist_from_stringfied_encoded_client(&self, detail: &StringfiedEncodedClient) -> anyhow::Result<()>{
+        let mut pool = self.pool.get().unwrap();
+        let client_str = serde_json::to_string(&detail)?;
+        pool.set(&detail.client_id, client_str)?;
+        Ok(())
+    }
+}
+
+impl OauthClientDBRepository for RedisDataSource{
+    fn list(&self, salt: String) -> anyhow::Result<Vec<EncodedClient>> {
+        unimplemented!()
```

Comment:
Is this unimplemented on purpose?

Reply:
Previously, I don't need the implement, and I think most users may not either. So I left to some user to impl as needed.
However,  my next version has impleted it.